jsg: dynamic imports

mikea opened the pull request:
Modules registered with addOnDemand method will be parsed and instantiated when all other imports failed.

mikea commented:
@jasnell @harrishancock please take a look.

jasnell commented on src/workerd/jsg/modules.h:
```diff
@@ -171,9 +171,9 @@ class ModuleRegistry {
   };
 
-  virtual kj::Maybe<ModuleInfo&> resolve(const kj::Path& specifier) = 0;
+  virtual kj::Maybe<ModuleInfo&> resolve(v8::Isolate* isolate, const kj::Path& specifier) = 0;
```
Let's have `resolve` take a `jsg::Lock& js` instead of the isolate

mikea replied:
Unfortunately that would mean changing very long call chains, since they don't have lock wired through.

E.g.:

resolve <- evaluateSyntheticModuleCallback <- createSyntheticModule <- ModuleInfo with no lock in sight yet.

Let me know if you want to create locks before calling resolve or tackle passing jsg::Lock everywhere in modules.h

mikea replied:
I'll make separate PR with this cleanup today if you don't mind.

jasnell commented on src/workerd/jsg/modules.h:
```diff
@@ -252,12 +260,29 @@ class ModuleRegistryImpl final: public ModuleRegistry {
+          return moduleInfo;
+        }
+        KJ_CASE_ONEOF(src, kj::ArrayPtr<const char>) {
+          info = ModuleInfo(isolate, specifier.basename()[0], src);
```
(to be clear, just saying that a comment here would be good :-) ...)

jasnell replied:
To import node core modules it should be `require('node:... ')` where `... ` is the name of the core module. E.g. `require('node:assert')`

mikea replied:
I don't think this name is relate to how you import. This name goes to ScriptOrigin.

I don't know if it should be the same or not.

PS nit: only ESM import will be supported, not require.

jasnell replied:
Why wouldn't require be supported? This should work for that also

mikea replied:
> Why wouldn't require be supported?

Because then we'll have to compile to version of the modules? One for ESM and one for commonjs?

Or can ESM modules be imported safely using require too?

jasnell replied:
The way we handle imports and require it should just work. Worth testing though

jasnell replied:
(to be clear, it *might* not just work also)

jasnell commented on src/workerd/jsg/modules.h:
```diff
@@ -198,17 +198,25 @@ class ModuleRegistryImpl final: public ModuleRegistry {
   }
 
-  kj::Maybe<ModuleInfo&> resolve(const kj::Path& specifier) override {
+  void addOnDemand(const kj::Path& specifier, kj::ArrayPtr<const char> sourceCode) {
```
`addOnDemand(...)` seems like an odd name for this. I would expect something more descriptive like `addBuiltinModule(...)`

Also, where and when would this be called? And by whom? I know this is part of a large piece of work but expanding the comment here to provide some additional detail would help me as a reviewer evaluate the correctness of this.

mikea replied:
I have created a demo in EW PR 4479. Will rename the function shortly.

mikea commented:
Method renamed, comments added. A change to replace Isolate with Lock in modules is underway. PTAL @jasnell

mikea commented:
> Method renamed, comments added. A change to replace Isolate with Lock in modules is underway. PTAL @jasnell

I think I messed up my various git clients. Hold on please while I fix this.

mikea commented:
pushed the update. To recap:

- renamed to `addBuiltinModule`
- added comments
- removed basename. Using `specifier.toString()` as module name.

ptal. @jasnell

kentonv commented on src/workerd/jsg/modules.h:
```diff
@@ -276,19 +305,28 @@ class ModuleRegistryImpl final: public ModuleRegistry {
+          return moduleInfo.hash;
+        }
+        KJ_CASE_ONEOF(src, kj::ArrayPtr<const char>) {
+          return kj::hashCode(src);
```
I guess this is hashing the whole source code, which sounds kind of expensive. Maybe we could key off a pointer to the code instead? This always points to a static constant, right?
